src/graphics/uniform-buffer-format.js

```diff
@@ -0,0 +1,107 @@
+import { Debug } from '../core/debug.js';
+import {
+    UNIFORMTYPE_BOOL, UNIFORMTYPE_INT, UNIFORMTYPE_FLOAT, UNIFORMTYPE_VEC2, UNIFORMTYPE_VEC3,
+    UNIFORMTYPE_VEC4, UNIFORMTYPE_IVEC2, UNIFORMTYPE_IVEC3, UNIFORMTYPE_IVEC4, UNIFORMTYPE_BVEC2,
+    UNIFORMTYPE_BVEC3, UNIFORMTYPE_BVEC4, UNIFORMTYPE_MAT4
+} from './constants.js';
+
+/** @typedef {import('./uniform-buffer.js').UniformBuffer} UniformBuffer */
+
+// map of UNIFORMTYPE_*** to byte size
+const uniformTypeToByteSize = [];
+uniformTypeToByteSize[UNIFORMTYPE_FLOAT] = 4;
+uniformTypeToByteSize[UNIFORMTYPE_VEC2] = 8;
+uniformTypeToByteSize[UNIFORMTYPE_VEC3] = 12;
+uniformTypeToByteSize[UNIFORMTYPE_VEC4] = 16;
+uniformTypeToByteSize[UNIFORMTYPE_INT] = 4;
+uniformTypeToByteSize[UNIFORMTYPE_IVEC2] = 8;
+uniformTypeToByteSize[UNIFORMTYPE_IVEC3] = 12;
+uniformTypeToByteSize[UNIFORMTYPE_IVEC4] = 16;
+uniformTypeToByteSize[UNIFORMTYPE_BOOL] = 4;
+uniformTypeToByteSize[UNIFORMTYPE_BVEC2] = 8;
+uniformTypeToByteSize[UNIFORMTYPE_BVEC3] = 12;
+uniformTypeToByteSize[UNIFORMTYPE_BVEC4] = 16;
+uniformTypeToByteSize[UNIFORMTYPE_MAT4] = 64;
+
+// Handle additiona types:
+//      UNIFORMTYPE_MAT2 = 12;
+//      UNIFORMTYPE_MAT3 = 13;
+//      UNIFORMTYPE_TEXTURE2D = 15;
+//      UNIFORMTYPE_TEXTURECUBE = 16;
+//      UNIFORMTYPE_FLOATARRAY = 17;
+//      UNIFORMTYPE_TEXTURE2D_SHADOW = 18;
+//      UNIFORMTYPE_TEXTURECUBE_SHADOW = 19;
+//      UNIFORMTYPE_TEXTURE3D = 20;
+//      UNIFORMTYPE_VEC2ARRAY = 21;
+//      UNIFORMTYPE_VEC3ARRAY = 22;
+//      UNIFORMTYPE_VEC4ARRAY = 23;
+
+/**
+ * A class storing description of an individual uniform, stored inside a uniform buffer.
+ *
+ * @ignore
+ */
+class UniformFormat {
```

Comment:
might make sense to add support for arrays early on because it will likely introduce another member here?

Reply:
yes I have a TODO comment somewhere there to do it, but at the moment the implementation has not been done. All this is private for now, so the interfaces can change as needed.